File: C++/1081-rational_sum.cc
#include <iostream>
#include <vector>
#include <string>
#include <numeric>


class Fraction {
public:
    Fraction() = default;
    Fraction(int numerator, int denominator);
    int numerator, denominator;
private:
    int _gcd(int a, int b);
};


Fraction::Fraction(int numerator, int denominator = 1) {
    this->numerator = numerator;
    this->denominator = denominator;
    if (numerator < 0) numerator *= -1;
    int factor = Fraction::_gcd(numerator, denominator);
    this->numerator /= factor;
    this->denominator /= factor;
}


int Fraction::_gcd(int a, int b) {
    while (b > 0) {
        int temp = a % b;
        a = b;
        b = temp;
    }
    return a;
}


Fraction operator+(const Fraction &f1, const Fraction &f2) {
    int numerator = f1.numerator * f2.denominator +
                    f2.numerator * f1.denominator;
    int denominator = f1.denominator * f2.denominator;
    return Fraction(numerator, denominator);
}


std::ostream &operator<<(std::ostream &os, const Fraction &f) {
    std::string sign;
    int numerator = f.numerator;
    if (numerator < 0) {
        sign = '-';
        numerator *= -1;
    }
    if (numerator != 0) {
        int integer_part = numerator / f.denominator;
        numerator %= f.denominator;
        if (integer_part != 0) {
            os << sign << integer_part;
            if (numerator != 0) {
                os << ' ' << numerator << '/' << f.denominator;
            }
        } else {
            os << sign << numerator << '/' << f.denominator;
        }
    } else {
        os << '0';
    }
    return os;
}


int main() {

    int n_fraction;
    std::cin >> n_fraction;
    std::vector<Fraction> fractions;
    while (n_fraction--) {
        int numerator, denominator;
        std::scanf("%d/%d", &numerator, &denominator);
        fractions.push_back(Fraction(numerator, denominator));
    }
    std::cout << std::accumulate(fractions.begin(), fractions.end(), Fraction(0));
    std::cout << '\n';

    return 0;
}

File: Python/1081-rational_sum.py


class Fraction():
    def __init__(self, numerator, denominator=1):
        self.numerator = numerator
        self.denominator = denominator
        factor = self._gcd(abs(numerator), denominator)
        if factor > 0:
            self.numerator //= factor
            self.denominator //= factor

    def __add__(self, other):
        numerator = (self.numerator * other.denominator +
                     other.numerator * self.denominator)
        denominator = self.denominator * other.denominator
        return Fraction(numerator, denominator)

    def _gcd(self, a, b):
        """gcd(a, b) = gcd(a % b, b) = gcd(b, a % b)"""
        while b > 0:
            a, b = b, a % b
        return a

    def __str__(self):
        sign = '' if self.numerator > 0 else '-'
        if self.numerator != 0:
            integer_part, numerator = divmod(abs(self.numerator),
                                             self.denominator)
            fraction_part = f'{numerator}/{self.denominator}'
            if integer_part != 0:
                res = sign + str(integer_part)
                if numerator != 0:
                    res += ' ' + fraction_part
            else:
                res = sign + fraction_part
        else:
            res = '0'
        return res


if __name__ == '__main__':
    _ = input()
    fractions = [Fraction(*[int(i) for i in deno_nume.split('/')])
                 for deno_nume in input().split()]
    print(sum(fractions, Fraction(0)))
